### ISSUE TITLE
weird interop with namespaces and JOI schemas

### ISSUE THREAD

**Issue by ssilve1989** [nestjs/config#914]
### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe it

When registering configurations via namespaces, like the following:
```typescript
const rabbitMQConfig = registerAs(
  'rabbitmq',
  () => ({
    hostname: process.env.RABBIT_HOST || 'localhost',
    port: process.env.RABBIT_PORT || 5672,
    user: process.env.RABBIT_USER,
    password: process.env.RABBIT_PASSWORD,
    timeout: process.env.RABBIT_TIMEOUT || 30000,
    wait: process.env.RABBIT_WAIT || false,
  })
);
```

and defining a validation schema as 
```typescript
    ConfigModule.forRoot({
      load: [rabbitMQConfig],
      validationSchema: Joi.object({
        RABBIT_HOST: Joi.string(),
        RABBIT_PORT: Joi.number(),
        RABBIT_USER: Joi.string().optional(),
        RABBIT_PASSWORD: Joi.string().optional(),
        RABBIT_TIMEOUT: Joi.number(),
        RABBIT_WAIT: Joi.boolean(),
      }),
    }),
```

using the type, as suggested in the documentation via `ConfigType<typeof rabbitMQConfig>` yields 
```typescript
{
    hostname: string;
    port: string | number;
    user: string | undefined;
    password: string | undefined;
    timeout: string | number;
    wait: string | boolean;
}
```

This creates a weird interaction with the validation schema, which already parses and transforms the value into the appropriate type. So `port: string | number;` is incorrect, since it will never be a string, due to the validationSchema being applied.

### Describe the solution you'd like

I'm not sure what solution to present here. I understand that namespaces function independently of a validation schema but it just creates this weird position to be in.

### What is the motivation / use case for changing the behavior?

To find some way to have accurate typing with the validations being applied by JOI schemas, without duplicating parsing/transforming logic inside the namespace function.

**Comment by micalevisk** [comment]
but:

```ts
const rabbitMQConfig = registerAs(
  'rabbitmq',
  () => {
   console.log(typeof process.env.RABBIT_PORT === 'string') // true
   return {
     hostname: process.env.RABBIT_HOST || 'localhost',
     port: process.env.RABBIT_PORT || 5672,
     user: process.env.RABBIT_USER,
     password: process.env.RABBIT_PASSWORD,
     timeout: process.env.RABBIT_TIMEOUT || 30000,
     wait: process.env.RABBIT_WAIT || false,
   }
  }
);
```

then excluding the union with `string` isn't right

**Comment by ssilve1989** [comment]
For some more clarification, it would seem that the namespace registration function just disregards the JOI conversion altogether. So what is the point in having a validation schema with a namespace if the namespace function is going to have to validate and convert it again for type-safety reasons? We lose the benefits of being able to express a Schema via JOI if we also want namespaces.

For example, assuming `FOO=100` is present on the environment and given that the `convert` validationOption for JOI defaults to true
```typescript
@Module({
  imports: [
    ConfigModule.forRoot({
      validationSchema: Joi.object({
        FOO: Joi.number().default(1),
      }),
    }),
  ],
})
export class SomeModule {
  constructor(private readonly configService: ConfigService) {
    console.log(typeof configService.get('FOO'), configService.get('FOO')); // number 100
  }
}
```
yields and converts types appropriately.

However if we want to continue expressing the schema that way, but additionally use a namespace
```typescript
const someConfig = registerAs('someConfig', () => {
  console.log(typeof process.env.FOO); // string
  return {
    foo: process.env.FOO || 1,
  };
});

@Module({
  imports: [
    ConfigModule.forRoot({
      validationSchema: Joi.object({
        FOO: Joi.number().default(1),
      }),
      load: [someConfig],
    }),
  ],
})
export class SomeModule {
  constructor(
    @Inject(someConfig.KEY)
    private readonly config: ConfigType<typeof someConfig>
  ) {
    console.log(typeof config.foo, config.foo); // string 100
  }
}

```
we've now rendered the Schema essentially useless, since to achieve what we want, we will still need to add number validation into the namespace function, as this for example:
```typescript
const someConfig = registerAs('someConfig', () => {
  console.log(typeof process.env.FOO); // string
  const foo = Number(process.env.FOO);
  return {
    foo: Number.isNaN(foo) ? 1 : foo,
  };
});
```

**Comment by ssilve1989** [comment]
I guess what I'm saying is the namespace function should be able to receive the validated/converted object if a validationSchema was specified? So it can choose to use values from there or the `process`

**Comment by micalevisk** [comment]
that would be a nice feature. I'm just not sure if it is feasible

**Comment by micalevisk** [comment]
The original Issue isn't an weird behavior, it's just how nodejs `process.env` works

```js
process.env.FOO = [1,2,3]
typeof process.env.FOO === 'string'
```

thus the type should have an union with `string` in your case. To circumvent this I've been using an utility function that will use the Joi schema and the env key so I always end up using the validated&parsed value instead of `process.env.`

**Comment by micalevisk** [comment]
> we've now rendered the Schema essentially useless ...

I believe the main issue around this scenario is that the factory you've supplied to `registerAs` will be register as a provider as-is (the `factory` function below):

https://github.com/nestjs/config/blob/247a8d608b485df5402fcabebe2f033f6d8723b7/lib/utils/create-config-factory.util.ts#L07-L15

while the `ConfigService.get` will look up at the parsed object

https://github.com/nestjs/config/blob/247a8d608b485df5402fcabebe2f033f6d8723b7/lib/config.service.ts#L103

---

But you can see that the validation step (below) happens before any call to `createConfigProvider`

https://github.com/nestjs/config/blob/fc2fedbe68559a1b47f9c41382b345364eaa3390/lib/config.module.ts#L69-L70

https://github.com/nestjs/config/blob/fc2fedbe68559a1b47f9c41382b345364eaa3390/lib/config.module.ts#L84

so I believe we can somehow pass the `validatedEnvConfig` object as the first arg of `registerAs`'s callbacks :thinking:

**Comment by micalevisk** [comment]
@ssilve1989

> the namespace function should be able to receive the validated/converted object

We can easily pass the parsed env. vars object as the first arg of `registerAs`'s factory, **but** when we use the [partial registration](https://docs.nestjs.com/techniques/configuration#partial-registration) feature (calling `.forFeature`), it will try to register that namespace function without having parsed the env vars yet (because this is done on `.forRoot` call).

https://github.com/nestjs/config/blob/fc2fedbe68559a1b47f9c41382b345364eaa3390/lib/config.module.ts#L142-L145

and I believe that is pretty much why this wasn't implemented.

So basically, the `ConfigFactory` callback will receive something only if we don't use the `.forFeature`, and I think this won't be good for DX :/

**Comment by micalevisk** [comment]
Also: [Ways to validate environment configuration in a forFeature Config in NestJs](https://dev.to/rrgt19/ways-to-validate-environment-configuration-in-a-forfeature-config-in-nestjs-2ehp).

---

I tried few things on `@nestjs/config` source. Now I believe we can manage to pass parsed env. vars. to the factories functions of `registerAs` so we don't need to validate those env. vars multiple times. So feel free to open another Issue to request such feature to see Kamil's thoughts as this is not related with this one.

**Comment by micalevisk** [comment]
btw I've made a POC for that feature here: https://gitlab.com/micalevisk/nestjs-config-new-feature

<details>
<summary>Results</summary>

```ts
import { ConfigModule, registerAs, ConfigType, ConfigService } from '@nestjs/config'
import { Module, Inject } from '@nestjs/common'
import * as Joi from 'joi'

interface Environment {
  FOO: number
}

const someConfig = registerAs('someConfig', (validatedEnvVars: Environment) => {
  console.log('[registerAs]', typeof process.env.FOO) // string
  console.log('[registerAs]', typeof validatedEnvVars.FOO) // number

  return {
    foo: process.env.FOO || 1, // *
  }
})

type SomeConfig = ConfigType<typeof someConfig>
/*
type SomeConfig = {
    foo: string | number;
}
*/

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      expandVariables: false,
      load: [someConfig],
      validationSchema: Joi.object({
        FOO: Joi.number().default(1), // *
      }),
      validationOptions: {
        allowUnknown: true,
        stripUnknown: true,
        abortEarly: true,
      },
    }),
  ],
})
export class AppModule {
  constructor(
    @Inject(someConfig.KEY)
    private readonly config: ConfigType<typeof someConfig>,
    private readonly configService: ConfigService,
  ) {}

  onModuleInit() {
    console.log('[onModuleInit]', typeof this.config.foo) // string
    console.log('[onModuleInit]', typeof this.configService.get('FOO')) // number
  }
}
```

</details>

**Comment by kamilmysliwiec** [comment]
@micalevisk thanks for your input!

> For some more clarification, the namespace registration function just disregards the JOI conversion altogether

This has been discussed many times in the past. `validationSchema` does not apply to custom namespace factories and there are no plans to change it in the foreseeable future.
